[5.5] Allow MailFake to assert mailables were queued

mateusjatenee opened the pull request:
Hey everyone,  

Right now when you queue a mailable you have to either assert it was `sent` (it was not, as it was queued) or you have to assert `Illuminate\Mail\SendQueuedMailable` was pushed on to the queue.  

This PR gives queued mailables their own property and allows assertions to be made, such as `assertWasQueued($mailable)`.    

Let me know what you guys think!

mateusjatenee committed:
implement mail fake queue method

mateusjatenee committed:
minor fixes

mateusjatenee committed:
Fix namespace declaration and change PendingMailFake queue implementation

mateusjatenee committed:
remove redundant class import

mateusjatenee committed:
proper docblock

mateusjatenee committed:
styleCI fix

thecrypticace commented on src/Illuminate/Support/Testing/Fakes/MailFake.php:
```diff
@@ -89,14 +137,50 @@ public function hasSent($mailable)
      * @return \Illuminate\Support\Collection
      */
-    protected function mailablesOf($type)
+    protected function mailablesOf($type, $queued = false)
```
Just my two cents but I'd make this a separate method. It would only be duplicating three lines and a method called `queuedMailablesOf` makes the intent much clearer.

mateusjatenee replied:
I agree, thank you!  

Added a commit that changes thi

mateusjatenee committed:
add queuedMailablesOf method and use the same implementation for send and queue

mateusjatenee commented:
Not entirely related, but perhaps we could build the view so the user knows that there are syntax errors or any other kind of error. I opened an issue in `laravel/internals` https://github.com/laravel/internals/issues/720

taylorotwell commented:
So would this be a breaking change? `assertSent` would no longer work for them?

taylorotwell commented:
@mateusjatenee building the view is a separate test IMO. and one that can already be written pretty easily.

taylorotwell commented:
This also doesn't work correctly if a Mailable is implementing ShouldQueue and is sent using Mail::send()

mateusjatenee committed:
Merge branch 'master' into feature/queud-mail-fake

mateusjatenee committed:
remove useless import

mateusjatenee committed:
fix hasQueued method and verify if the mailable should be queued on send method

mateusjatenee committed:
add assertQueuedTimes method

mateusjatenee commented:
@taylorotwell Sorry I forgot to queue it if It implemented `ShouldQueue` — I did this late and forgot it.  

I added some code that verifies if the mailable implements the interface and queues if it does.  Given https://github.com/laravel/framework/pull/20485 was approved I also added the option to specify how many times it should have been queued.  

It would be a BC, yes.. but I believe it's worth it as the "migration" is very simple and verifying a mailable was queued with `sent` doesn't really make sense I guess.

mateusjatenee committed:
Merge branch 'master' into feature/queud-mail-fake

mateusjatenee committed:
add tests

mateusjatenee committed:
test a mailable that implements ShouldQueue is queued and not sent

mateusjatenee committed:
styleCI fixes

mateusjatenee committed:
merge master and change exception checks to annotations

taylorotwell commented:
Thanks
